appscale/tools/appcontroller_client.py

```diff
@@ -227,6 +227,46 @@ def start_roles_on_nodes(self, roles_to_nodes):
     return self.run_with_timeout(self.DEFAULT_TIMEOUT, "Error", self.DEFAULT_NUM_RETRIES,
       self.server.start_roles_on_nodes, roles_to_nodes, self.secret)
 
+  def is_appscale_terminated(self):
+    """Dynamically adds the given machines to an AppScale deployment, with the
+    specified roles.
+
+    Returns:
+      A boolean indicating whether appscale has finished running terminate
+        on all nodes.
+    """
+    return self.run_with_timeout(self.DEFAULT_TIMEOUT, "Error",
+                                 self.DEFAULT_NUM_RETRIES,
+                                 self.server.is_appscale_terminated, self.secret)
+
+  def run_terminate(self, clean):
+    """Dynamically adds the given machines to an AppScale deployment, with the
+    specified roles.
+
+    Args:
+      clean: A boolean indicating whether the clean parameter should be
+        passed to terminate.rb.
+    Returns:
+      The result of executing the SOAP call on the remote AppController.
+    """
+    return self.server.run_terminate(clean, self.secret)
+
+
+  def receive_server_message(self):
+    """Dynamically adds the given machines to an AppScale deployment, with the
+    specified roles.
+
+    Args:
+      clean: A boolean indicating whether the clean parameter should be
+        passed to terminate.rb.
+    Returns:
+      The result of executing the SOAP call on the remote AppController.
+    """
+    return self.run_with_timeout(self.DEFAULT_TIMEOUT * 5, "Error",
+                                 self.DEFAULT_NUM_RETRIES,
+                                 self.server.receive_server_message,
+                                 self.DEFAULT_TIMEOUT * 4,
```

Comment:
The server side needs to also have a timeout so that it doesn't clear the queue of messages because it thinks the client got them. I made the server timeout smaller than the client timeout so that there wouldn't be any issues with it. Without the timeout on the server side the client would run the method and then time out possibly before the server returns a message, however the server won't know the client timed out and will assume the message is received and clear the message.